src/Sulu/Bundle/ContentBundle/Controller/ContentController.php

```diff
@@ -0,0 +1,347 @@
+    {
+        $properties = array_filter(explode(',', $request->get('mapping', '')));
+        $excludeGhosts = $this->getBooleanRequestParameter($request, 'exclude-ghosts', false, false);
+        $excludeShadows = $this->getBooleanRequestParameter($request, 'exclude-shadows', false, false);
+        $webspaceNodes = $this->getRequestParameter($request, 'webspace-nodes');
+        $tree = $this->getBooleanRequestParameter($request, 'tree', false, false);
+        $locale = $this->getRequestParameter($request, 'locale', true);
+        $webspaceKey = $this->getRequestParameter($request, 'webspace', true);
+
+        $user = $this->tokenStorage->getToken()->getUser();
+
+        $mapping = MappingBuilder::create()
+            ->disableHydrateGhost(!$excludeGhosts)
+            ->disableHydrateShadow(!$excludeShadows)
+            ->addProperties($properties)
+            ->getMapping();
+
+        if ($tree) {
+            return $this->getTreeAction($uuid, $locale, $webspaceKey, $webspaceNodes, $mapping, $user);
+        }
+
+        $data = $this->contentRepository->find($uuid, $locale, $webspaceKey, $mapping, $user);
+        $view = $this->view($data);
+
+        return $this->viewHandler->handle($view);
+    }
+
+    /**
+     * Returns tree response for given uuid.
+     *
+     * @param string $uuid
+     * @param string $locale
+     * @param string $webspaceKey
+     * @param bool $webspaceNodes
+     * @param MappingInterface $mapping
+     * @param UserInterface $user
+     *
+     * @return Response
+     *
+     * @throws ParameterDataTypeException
+     */
+    private function getTreeAction(
+        $uuid,
+        $locale,
+        $webspaceKey,
+        $webspaceNodes,
+        MappingInterface $mapping,
+        UserInterface $user
+    ) {
+        if (!in_array($webspaceNodes, [self::WEBSPACE_NODE_SINGLE, self::WEBSPACE_NODES_ALL, null])) {
+            throw new ParameterDataTypeException(get_class($this), 'webspace-nodes');
+        }
+
+        try {
+            $contents = $this->contentRepository->findParentsWithSiblingsByUuid(
+                $uuid,
+                $locale,
+                $webspaceKey,
+                $mapping,
+                $user
+            );
+        } catch (ItemNotFoundException $ex) {
+            // TODO return 404 and handle this edge case on client side
+            return $this->redirect(
+                $this->router->generate(
+                    'get_contents',
+                    [
+                        'locale' => $locale,
+                        'webspace' => $webspaceKey,
+                        'exclude-ghosts' => !$mapping->hydrateGhost(),
+                        'exclude-shadows' => !$mapping->hydrateShadow(),
+                        'mapping' => implode(',', $mapping->getProperties()),
+                    ]
+                )
+            );
+        }
+
+        if ($webspaceNodes === self::WEBSPACE_NODES_ALL) {
```

Comment:
Please use `static::` to access constants, because the class is easier to extend then. (also all other appearances of `self::`)